packages/@react-stately/combobox/src/useComboBoxState.ts

```diff
@@ -0,0 +1,304 @@
+  }
+
+  getFirstKey() {
+    return this.firstKey;
+  }
+
+  getLastKey() {
+    return this.lastKey;
+  }
+
+  getItem(key: Key) {
+    return this.keyMap.get(key);
+  }
+}
+
+let whitespace = /\s/;
+
+export function useComboBoxState<T extends object>(props: ComboBoxProps<T>): ComboBoxState<T> {
+  let {
+    onFilter,
+    selectedKey,
+    defaultSelectedKey,
+    inputValue,
+    defaultInputValue,
+    onInputChange,
+    isFocused,
+    isOpen,
+    defaultOpen
+  } = props;
+  
+  let itemsControlled = !!onFilter;
+  let collator = useCollator({sensitivity: 'base'});
+  /*
+  let menuControlled = props.isOpen !== undefined;
+  let valueControlled = inputValue !== undefined;
+  let selectedControlled = !!selectedKey;
+   */
+
+  // Create a separate menu open state tracker so onOpenChange doesn't fire with open and close in quick succession
+  // in cases where there aren't items to show
+  // Note that this means onOpenChange won't fire for controlled open states
+  let [menuIsOpen, setMenuIsOpen] = useControlledState(isOpen, defaultOpen || false, () => {});
+
+  let selectState = useSelectState(props);
+
+  let selectedKeyItem = selectedKey ? selectState.collection.getItem(selectedKey) : undefined;
+  let selectedKeyText = selectedKeyItem ? selectedKeyItem.textValue || selectedKeyItem.rendered as string : undefined;
+  // Maybe don't need to do for defaultSelectedKey? Need to do for selectedKey so that the textfield is properly controlled reflects selectedKey text
+  let defaultSelectedKeyItem = defaultSelectedKey ? selectState.collection.getItem(defaultSelectedKey) : undefined;
+  let defaultSelectedKeyText = defaultSelectedKeyItem ? defaultSelectedKeyItem.textValue || defaultSelectedKeyItem.rendered as string : undefined;
+  // Double check if selectedKey should make textfield value controlled
+  let [value, setValue] = useControlledState(toString(inputValue) || selectedKeyText, toString(defaultInputValue) || defaultSelectedKeyText || '', onInputChange);
+  let lowercaseValue = value.toLowerCase();
+
+  let defaultFilterFn = useMemo(() => (node: Node<T>) => {
+    let scan = 0;
+    let lowercaseNode = node.textValue.toLowerCase();
+    for (let i in lowercaseValue) {
+      if (whitespace.test(lowercaseValue[i])) {
+        continue;
+      }
+      let match = false;
+      for (; scan < lowercaseNode.length && !match; scan++) {
+        if (!whitespace.test(lowercaseNode[scan])) { // if whitespace, move on to next character
+          let compareVal = collator.compare(lowercaseValue[i], lowercaseNode[scan]);
+          if (compareVal !== 0) { // if it didn't match, we can exit
+            return false;
+          } else { // we matched, move to next character in input for comparison
+            match = true;
+          }
+        }
+      }
+    }
+    return true;
+  }, [collator, lowercaseValue]);
+
+  useEffect(() => {
+    if (onFilter) {
+      onFilter(value);
+    }
+  // Having onFilter in the dep array seems to break it
```

Comment:
🤔 I think it's correct though. If it's not in there, and the `onFilter` function changes, then we'd be calling the wrong one later on. Perhaps you could specifically compare whether the value changed with a ref to track the last value:

```js
let lastValue = useRef('');
useEffect(() => {
  if (onFilter && lastValue.current !== value) {
    onFilter(value);
  }

  lastValue.current = value;
}, [value, onFilter])